PythonVisualizations/Heap.py

```diff
@@ -0,0 +1,518 @@
+        # if the array is now empty, delete the NoneType place holder from the array
+        if self.size < 1: del self.list[:]
+        # Slide value rectangle up and off screen
+        items = (n.display_shape, n.display_val)
+        toPositions = (100, -100, 100, -50)
+        callEnviron |= set(items)
+        self.moveItemsBy(items, delta=(0, -max(400, self.canvas.coords(n.display_shape)[3])), steps=self.CELL_HEIGHT,
+                        sleepTime=.05)
+
+        if self.size < 1:
+            self.setMessage("Heap requirement satisfied")
+
+        # move bottom cell to top, and heapify
+        else:
+            self.swapRoot()
+            self.heapify()
+
+        # Finish animation
+        self.cleanUp(callEnviron)
+
+
+    def isHeap(self):
+        cur = 0
+        size = len(self.list)
+
+        while cur < size // 2:
+            leftChild = 2 * cur + 1
+            rightChild = leftChild + 1
+
+            top = self.list[cur]
+            if self.list[leftChild] > top or (rightChild < size and self.list[rightChild] > top):
+                return False  # return false if a child is larger than a parent
+            else:
+                cur += 1
+
+        return True
+
+
+    def heapify(self):
+        # trickle down until heap is in order
+        while not self.isHeap():
+            self.trickleDown()
+
+
+    def swapRoot(self):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        n = self.list[-1]
+        items = (n.display_shape, n.display_val)
+
+        # move the last cell to the front of the heap
+        cellCoords = self.cellCoords2(0)
+        cellCenter = self.cellCenter2(0)
+        toPositions = (cellCoords, cellCenter)
+        # self.moveItemsOffCanvas(items, N, sleepTime=0.02) ## alternative method to moveItemsBy
+        self.moveItemsTo(items, toPositions)
+        self.list[0] = n
+        # delete the last element from the list
+        del self.list[-1]
+
+
+    def trickleDown(self, cur=0, size=-1):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+
+        if size == -1: size = len(self.list)
+        top = self.list[cur]  # save the Node at the root
+
+        # while the current Node has at least one child...
+        while cur < size // 2:
+            leftChild = 2 * cur + 1
+            rightChild = leftChild + 1
+            largerChild = leftChild
+            top = self.list[cur]
+            # find smaller child (right child might not exist)
+            if rightChild < size and \
+                    self.list[leftChild] < self.list[rightChild]:
+                largerChild = rightChild
+            # done trickling if top's key is >= the key of larger child
+            if top >= self.list[largerChild]:
+                cur += 1
+
+            else:
+                # shift child up
+                self.swap(cur, largerChild)
+                cur += 1
+
+        self.cleanUp(callEnviron)

```

Comment:
Use the `siftDown()` algorithm from the textbook code.   This one always visits the first half of the heap array.  The sift-down operation in the textbook code only follows one path down from the root, doesn't use swap, and finishes as soon as the item sifting down is in correct position.